# Лабораторная работа 1. Язык Python, основные библиотеки для анализа данных.

ФИО: Гурьянов Алексей Константинович

Группа: 517

In [1]:
import numpy as np
import pandas

## Часть 1: Python

#### 1. Напишите код, который сформирует список всех натуральных чисел, не превосходящих 100, которые делятся на 7 и не делятся на 5. Нельзя использовать циклы — выполните это задание с помощью list comprehension.

In [2]:
 [x for x in range(1, 101) if x % 7 == 0 and x % 5 != 0]

[7, 14, 21, 28, 42, 49, 56, 63, 77, 84, 91, 98]

#### 2. Напишите функцию `sort_tokens(s)`, которая принимает на вход строку, разделяет ее на токены по запятым, сортирует токены с помощью функции `sorted()` и снова соединяет их в одну строку через запятые. Например, для строки `u"ночь,фонарь,улица"` функция должна выдать `u"ночь,улица,фонарь"`. Придумайте несколько входных строк и протестируйте функцию на них.

Обратите внимание, что склеить строки можно с помощью конструкции `'x'.join(...)`. Например:

In [3]:
'! '.join(['i', 'am', 'very', 'excited', ':-)'])

'i! am! very! excited! :-)'

In [4]:
def sort_tokens(s):
    split_s = s.split(',')
    sorted_split_s = sorted(split_s)
    return ','.join(sorted_split_s)

#### 3. Напишите функцию `remove_duplicates(a)`, которая принимает на вход список и возвращает его же, но без дубликатов и в отсортированном виде. Например, для списка `[u"ночь", u"улица", u"фонарь", u"аптека", u"аптека", u"улица", u"фонарь"]` результат должен быть `[u"аптека", u"ночь", u"улица", u"фонарь", ]`.

Подсказка: обратите внимание на структуру `set`!

In [5]:
def remove_duplicates(a):
    return sorted([x for x in set(a)])

#### 4. Напишите функцию `word_counts(texts)`, которая принимает на вход список строк, и печатает пары (слово, число вхождений). Считайте, что слова в текстах могут быть разделены только пробелами. Например, для входа `[u'ночь улица фонарь аптека', u'аптека улица фонарь']` должно быть напечатано
>аптека 2

>улица 2

>фонарь 2

>ночь 1

Подсказки:

Чтобы разбить каждый из текстов на слова, воспользуйтесь функцией `map`, которая применяет заданную функцию к каждому элементу списка. Например, если бы нужно было каждый из текстов привести к нижнему регистру, то команда выглядела бы так:

In [6]:
map(lambda x: x.lower(), ["abCDefg", "HIjkL"])

['abcdefg', 'hijkl']

После первого шага вы получите список, состоящий из списков слов. Чтобы объединить их в один список слов, воспользуйтесь функцией `reduce`. Например, просуммировать числа из списка с ее помощью можно так:



In [7]:
reduce(lambda x, y: x + y, [1, 2, 3, 4, 5])

15

Для подсчета количества вхождений слов воспользуйтесь структурой `defaultdict` из модуля `collections`. В отличие от обычного `dict`, она не генерирует ошибку при обращении к несуществущему ключу, а создает новую запись, инициализируемую нулевым значением. Например:

In [8]:
from collections import defaultdict
d = defaultdict(int)
print d
d['abc'] += 1
print d
d['xxx']
print d

defaultdict(<type 'int'>, {})
defaultdict(<type 'int'>, {'abc': 1})
defaultdict(<type 'int'>, {'xxx': 0, 'abc': 1})


In [9]:
from collections import defaultdict

def compute(x, y):
    for word in y:
        x[word] += 1
        
    return x

def word_counts(texts):
    split_texts = map(lambda x: x.split(' '), texts)
    d = defaultdict(int)
    return reduce(lambda x, y: compute(x, y) , [d] + split_texts)    

#### 5. (бонусное задание) Скачайте английский перевод "Преступления и наказания". Найдите 20 самых популярных слов и 20 слов, которые встречаются только один раз.

Приведите текст к нижнему регистру. Разделите текст на слова, считая разделителями все, кроме букв. Обратите внимание на функцию `split` из модуля `re`.

Не забудьте удалить пустые слова:
```
words = filter(lambda x: len(x) > 0, words)
```

In [10]:
from urllib import urlopen
url = "http://www.gutenberg.org/files/2554/2554.txt"
response = urlopen(url)
raw_text = response.read().decode('utf8')

In [11]:
r = raw_text[1000:1100]

In [12]:
lowered_raw = raw_text.lower()

In [13]:
import re

In [14]:
split_text = re.split('[^a-z]*', lowered_raw )

In [15]:
words = filter(lambda x: len(x) > 0, split_text)

In [16]:
d = defaultdict(int)
for word in words:
    d[word] += 1

In [17]:
sorted_words = sorted(d.items(), key = lambda x: x[1], reverse = True)
frequent_20 = [x[0] for x in sorted_words[0:20]]

one_timers = filter(lambda x: x[1] == 1, sorted_words)
one_timers_20 = [x[0] for x in one_timers[0:20]]

20 самых часто употребляемых слов:

In [18]:
print (frequent_20)

[u'the', u'and', u'to', u'he', u'a', u'i', u'you', u'of', u'it', u'that', u'in', u'was', u'his', u'at', u'her', u'but', u'not', u's', u'with', u'she']


20 слов, которые встречались один раз:

In [19]:
print (one_timers_20)

[u'fawn', u'writings', u'nunnery', u'fedosya', u'clotted', u'preface', u'cyprus', u'increase', u'wheeled', u'superficially', u'meadows', u'fogeys', u'anaemic', u'straws', u'busybody', u'reliable', u'scraped', u'errors', u'cooking', u'contributed']


## Часть 2: NumPy и SciPy

#### 1. Повторите 1000 раз следующий эксперимент: сгенерируйте две матрицы размера $10 \times 10$ из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

Для вычисления квантили воспользуйтесь функцией `np.percentile`

In [20]:
import numpy as np

max_elements = []

for i in range (0, 1000):
    a = np.matrix(np.random.normal(0, 1, (10, 10)))
    b = np.matrix(np.random.normal(0, 1, (10, 10)))
    c = a * b
    max_elements.append(c.max())
    
max_average = np.average(max_elements)
max_95q = np.percentile(max_elements, 95)

print (max_average, max_95q)

(8.4732193319955833, 11.935489288005311)


#### 2. Сгенерируйте 1000 чисел из распределения Пуассона с параметром $\lambda = 5$. Сформируйте массив, в котором в i-м элементе будет записано количество сгенерированных чисел, равных i.

Обратите внимание на функцию `np.bincount`

In [21]:
poiss = np.random.poisson(5, 1000)

binCounts = np.bincount(poiss)
print (binCounts)

[  2  45  81 129 187 151 135 116  64  46  23  11   5   2   3]


#### 3. Реализуйте функцию, которая принимает на вход numpy-массив целых чисел `a`, и генерирует массив, в котором число `i` встречается `a[i]` раз. Протестируйте на результате из прошлого задания.

Обратите внимание на функцию `np.repeat`

In [22]:
def allRepeat(a):
    return np.repeat(np.arange(0, len(a)), a)
    
restored = allRepeat(binCounts)
sorted_poiss = np.sort(poiss)

if np.array_equal(restored, sorted_poiss):
    print('Test succeeded')

Test succeeded


#### 4. Напишите функцию `scale(X)`, которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на каких-нибудь данных.

In [23]:
def func(row):
    average = np.average(row)
    std = np.std(row)
    if std == 0:
        return [0] * len(row)
    else:
        return [float(x - average) / std for x in row]

def scale(X):
    return np.matrix(np.apply_along_axis(func, axis = 0, arr = X))

a = np.matrix([[1,2,3],[3,4,5],[2,3,4],[4,5,6]])
print (a)
s1 = scale(a)
print (s1)

b = np.matrix([[3,2],[4,2],[1,2],[4,2]])
print (b)
s2 = scale(b)
print (s2)


[[1 2 3]
 [3 4 5]
 [2 3 4]
 [4 5 6]]
[[-1.34164079 -1.34164079 -1.34164079]
 [ 0.4472136   0.4472136   0.4472136 ]
 [-0.4472136  -0.4472136  -0.4472136 ]
 [ 1.34164079  1.34164079  1.34164079]]
[[3 2]
 [4 2]
 [1 2]
 [4 2]]
[[ 0.          0.        ]
 [ 0.81649658  0.        ]
 [-1.63299316  0.        ]
 [ 0.81649658  0.        ]]


#### 5. Сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}(10, 1)$. Найдите ее:
- определитель
- след
- наименьший и наибольший элементы
- спектральную норму
- норму Фробениуса
- собственные числа
- обратную матрицу

Генерация произвольной матрицы

In [24]:
mat = np.matrix(np.random.normal(10, 1, (5,5)))
print(mat)

[[  8.40584489   9.74559626  10.17286135   9.82618748  11.23873353]
 [ 10.92291767  12.1155741    9.52090089  10.86909363  10.94854595]
 [ 10.94677303   8.46407664  10.44622341   9.40245916  10.08852451]
 [  9.52630783   9.53326169  10.52746286   9.86276184   9.84049195]
 [  9.73050665  10.72485216  11.15138563  10.40401366   9.49351727]]


Вычисление определителя

In [25]:
det = np.linalg.det(mat)
print (det)

42.9859722397


Вычисление следа

In [26]:
trace = mat.trace()
print (trace)

[[ 50.32392151]]


Вычисление минимального и максимального элемента

In [27]:
max = np.max(mat)
min = np.min(mat)
print (max)
print (min)

12.1155741041
8.40584488996


Вычисление спектральной нормы

In [28]:
spectr_norm = np.linalg.norm(mat, ord = 2)
print(spectr_norm)

50.8266295336


Вычисление нормы Фробениуса

In [29]:
fro_norm = np.linalg.norm(mat, ord = "fro")
print(fro_norm)

50.956037549


Вычисление собственных значений

In [30]:
eigenval, eigenvec = np.linalg.eig(mat)
print (eigenval)

[ 50.78399740+0.j          2.32066417+0.j         -1.44827126+1.0257752j
  -1.44827126-1.0257752j   0.11580246+0.j       ]


Вычисление обратной матрицы

In [31]:
inverse = np.linalg.inv(mat)
print (inverse)
Id = inverse * mat
print (Id)

[[-0.11498407  0.11639341  0.4785153  -0.69060273  0.20922635]
 [ 0.71564763  0.06660437  0.69494299 -3.26278614  1.71951838]
 [ 0.56062362 -0.40596545  0.65453156 -2.19207287  1.38113717]
 [-1.78749423  0.26925081 -2.2175952   7.39505462 -3.50317428]
 [ 0.6097869  -0.01275615  0.38590541 -1.13559449  0.16516755]]
[[  1.00000000e+00  -9.76996262e-15  -1.28785871e-14  -1.11022302e-14
   -1.15463195e-14]
 [  3.55271368e-15   1.00000000e+00   3.55271368e-15   1.06581410e-14
   -5.32907052e-15]
 [ -5.32907052e-15  -7.10542736e-15   1.00000000e+00  -5.32907052e-15
   -8.88178420e-15]
 [  2.13162821e-14   1.42108547e-14   1.42108547e-14   1.00000000e+00
    1.42108547e-14]
 [  1.33226763e-15  -1.11022302e-15   1.11022302e-15   1.55431223e-15
    1.00000000e+00]]


## Часть 3: Pandas

#### Ответьте на вопросы о данных по авиарейсам в США за январь-апрель 2008 года.

Данные: http://stat-computing.org/dataexpo/2009/2008.csv.bz2
(обратите внимание, что распаковывать этот файл не обязательно — функция `pandas.read_csv` умеет читать из архивов автоматически)

Описание: http://stat-computing.org/dataexpo/2009/the-data.html

1. Какая из причин отмены рейса (`CancellationCode`) была самой частой? (расшифровки кодов можно найти в описании данных)
2. Найдите среднее, минимальное и максимальное расстояние, пройденное самолетом.
3. Не выглядит ли подозрительным минимальное пройденное расстояние? В какие дни и на каких рейсах оно было? Какое расстояние было пройдено этими же рейсами в другие дни?
4. Из какого аэропорта было произведено больше всего вылетов? В каком городе он находится?
5. Найдите для каждого аэропорта среднее время полета (`AirTime`) по всем вылетевшим из него рейсам. Какой аэропорт имеет наибольшее значение этого показателя?
6. Найдите аэропорт, у которого наибольшая доля задержанных (`DepDelay > 0`) рейсов. Исключите при этом из рассмотрения аэропорты, из которых было отправлено меньше 1000 рейсов (используйте функцию `filter` после `groupby`).

In [32]:
df = pandas.read_csv('2008.csv.bz2')

In [33]:
df.CancellationCode.describe()

count     137434
unique         4
top            B
freq       54904
Name: CancellationCode, dtype: object

Самая частая причина отмены рейса имела код B - это погода. В принципе логично.

In [34]:
df['Distance'].mean()

726.38702942539283

In [35]:
df['Distance'].max()

4962

In [36]:
df['Distance'].min()

11

In [37]:
distance11 = df.query("Distance==11")
distance11

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
2547298  2008      5          15          4     2037        1800     2125   
4392215  2008      8          10          7     1315        1220     1415   

         CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  \
2547298        1900            OH       4988        ...             10   
4392215        1320            OH       5572        ...              8   

         TaxiOut  Cancelled  CancellationCode  Diverted  CarrierDelay  \
2547298        7          0               NaN         0           145   
4392215       38          0               NaN         0            55   

        WeatherDelay NASDelay  SecurityDelay  LateAircraftDelay  
2547298            0        0              0                  0  
4392215            0        0              0                  0  

[2 rows x 29 columns]

In [38]:
print (distance11.Origin)
print (distance11.Dest)

2547298    JFK
4392215    JFK
Name: Origin, dtype: object
2547298    LGA
4392215    LGA
Name: Dest, dtype: object


Код аэропорта JFK соответствует международный аэропорт Джонна Кеннеди в Нью Йорке(координаты 40.63975111,-73.77892556), код LGA соответствует аэропорту La Guardia в Нью Йорке(координаты - 40.77724306,-73.87260917)

Самолеты пролетали 11 миль два дня во всей выборке, 15 мая 2008(рейс 4988 JFK-LGA) и 10 августа 2008 (рейс 5572 JFK-LGA)

In [49]:
df[df.Origin=="JFK"][df.Dest=="LGA"]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
2547298  2008      5          15          4     2037        1800     2125   
4392215  2008      8          10          7     1315        1220     1415   

         CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  \
2547298        1900            OH       4988        ...             10   
4392215        1320            OH       5572        ...              8   

         TaxiOut  Cancelled  CancellationCode  Diverted  CarrierDelay  \
2547298        7          0               NaN         0           145   
4392215       38          0               NaN         0            55   

        WeatherDelay NASDelay  SecurityDelay  LateAircraftDelay  
2547298            0        0              0                  0  
4392215            0        0              0                  0  

[2 rows x 29 columns]

Больше не было ни одного рейса из первого аэропорта во второй.

In [40]:
df.Origin.describe()

count     7009728
unique        303
top           ATL
freq       414513
Name: Origin, dtype: object

Чаще всего самолеты вылетали из аэропорта ATL, аэропорта Atlantic City.

In [41]:
mean_airtimes = df.groupby("Origin").AirTime.mean()

In [42]:
max_mean_port = mean_airtimes.idxmax()
max_mean = mean_airtimes[max_mean_port]

print (max_mean_port, max_mean)

('SJU', 203.8508426966292)


Максимальная средняя длительность полета достигается в аэропорте SJU - аэропорт Сан-Хуан, столицы Пуэрто-Рико.

In [43]:
grouped = df.groupby("Origin")
df_trunc = grouped.filter(lambda x: len(x) >= 1000)

In [47]:
trunc_grouped = df_trunc.groupby("Origin")
stats = trunc_grouped.apply(lambda x: float(len(np.where(np.array(x) > 0))) / len(x) )

In [48]:
print (stats.idxmax())

LCH


Итого аэропорт с максимальной долей отложенных вылетов среди аэропортов с более чем 1000 вылетами - это аэропорт Lake Charles Regional в Луизиане.